Pod/Classes/LSBinaryCodec.m

```diff
@@ -0,0 +1,324 @@
+
+    switch (keyInfo.fieldType) {
+        case WireFormatVarint:
+            keyInfo.length = 0;
+            break;
+        case WireFormatFixed64:
+            keyInfo.length = sizeof(UInt64);
+            break;
+        case WireFormatFixed32:
+            keyInfo.length = sizeof(UInt32);
+            break;
+        case WireFormatLengthDelim:
+        {
+            UInt64 length = [LSBinaryCodec readVarintFromProto:protoEnc startingAt:offset];
+            keyInfo.length = length;
+            break;
+        }
+        default:
+            break;
+    }
+
+    return keyInfo;
+}
+
+// Reads a varint off the proto bytes, advances the offset to the next byte after the varint.
++ (UInt64)readVarintFromProto:(NSData *)proto startingAt:(UInt64 *)offset {
+    UInt64 varint = 0;
+    UInt32 shift = 0;
+    while (shift < 64) {
+        // Read a byte.
+        Byte *datum = (Byte *)proto.bytes + (*offset * sizeof(Byte));
+        *offset += 1;
+
+        varint |= (*datum & 0x7F) << shift;
+        shift += 7;
+
+        // If MSB was not set, we're done.
+        if ((*datum & 0x80) == 0) {
+            return varint;
+        }
+
+        // Now you have a uint!
+    }
+
+    return 0;
+}
+
++ (void)writeTagNumber:(UInt64)tagNumber format:(Byte)format buffer:(NSMutableData *)buffer {
+    UInt64 value = tagNumber << 3 | format;
+    [LSBinaryCodec writeVarintEncodedUInt64:value buffer:buffer];
+}
+
++ (void)writeStringMap:(NSDictionary *)data tagNumber:(UInt64)tagNumber buffer:(NSMutableData *)buffer {
+    for (id k in data) {
+        id v = [data objectForKey:k];
+
+        // TODO: Try to avoid this extra alloc?
+        NSMutableData *oneEntry = [[NSMutableData alloc] init];
+
+        // for key & value
+        // Write the equivalent tag key.
+        // Write the length.
+        // Write the string bytes.
+
+        NSString *key = (NSString *)k;
+        NSUInteger kl = [key lengthOfBytesUsingEncoding:NSUTF8StringEncoding];
+        [LSBinaryCodec writeTagNumber:1 format:WireFormatString buffer:oneEntry];
+        [LSBinaryCodec writeVarintEncodedUInt64:kl buffer:oneEntry];
+        [oneEntry appendData:[key dataUsingEncoding:NSUTF8StringEncoding]];
+
+        NSString *value = (NSString *)v;
+        NSUInteger vl = [value lengthOfBytesUsingEncoding:NSUTF8StringEncoding];
+        [LSBinaryCodec writeTagNumber:2 format:WireFormatString buffer:oneEntry];
+        [LSBinaryCodec writeVarintEncodedUInt64:vl buffer:oneEntry];
+        [oneEntry appendData:[value dataUsingEncoding:NSUTF8StringEncoding]];
+
+        // Write the message tag + type.
+        [LSBinaryCodec writeTagNumber:tagNumber format:WireFormatMap buffer:buffer];
+
+        // Write the total message length
+        [LSBinaryCodec writeVarintEncodedUInt64:oneEntry.length buffer:buffer];
+
+        // Append the message itself
+        [buffer appendData:oneEntry];
+    }
+}
+
++ (void)writeFixedEncodedUInt64:(UInt64)number buffer:(NSMutableData *)buffer {
```

Comment:
it would be easier to grok/review if we move the truly generic stuff (like this) into its own compilation unit